### PR TITLE
[Middleware_01] Document Internal Class Implementation

### LINKED ISSUE #40: [Middleware_01] Document Internal Class Implementation
**What is the use case? Please describe.**
Document the internal implementation of classes in the `edu.oakland.production.middleware01` or `edu.oakland.helper.middleware01` packages. Note that this is part of sprint deliverable tasks 1.4, 1.5, 2.1.4, 2.1.5, and 2.3.2.

**Sprint Task Setup**

To be done by @brendam19 , @juliawang2014 , or @martinhughes0722 before the rest of the team starts working on it:

1. Use the documentation at https://github.com/andrewdimmer/csi-3370-software-project/tree/middleware_01-production-uml-class-diagrams/docs/uml-class-diagrams/middleware01 to identify the classes and `README.md` files to update.
2. Edit this issue message to add a checkbox for each class, and assign it to the correct team member(s).

**Software Setup**

For this task, you'll be updating PlantUML diagrams. To preview them on your local machine, set up:

- [Visual Studio Code](https://code.visualstudio.com/)
- [PlantUML Visual Studio Code Extension](https://marketplace.visualstudio.com/items?itemName=jebbs.plantuml)
- [Graphviz](https://graphviz.org/)

After that, you can open a .puml document, right click on it, and "Preview Current Diagram".

If you don't want to set that up though, you don't have to. When you commit, GitHub has been configured to automatically update all the diagrams, and you can push to GitHub then view the diagrams from there and/or pull them back to your current machine.

**Steps**

To be done by each team member for each class/interface or package they are documenting:

1. Checkout and/or pull the branch `middleware_01-production-uml-class-diagrams`
2. Open the docs in the `/docs/uml-class-diagrams/middleware01` folder.
3. Navigate via the links in the `README.md` (on GitHub to view) or the file structure (on your local machine to edit) to the class you're updating (or the package if you are updating the package README.md files).
4. Edit the correct `README.md` document, and/or the document titled `NAME_OF_CLASS.puml`. Need help with the PlantUML Diagram itself? Check out https://plantuml.com/class-diagram.
5. Commit your changes with the message `[Middleware_01] Added Class/Interface_Name_Here UML Class Diagram`
6. Push your changes
7. If you are the first person to push, create a pull request draft with the name `[Middleware_01] Document Internal Class Implementation` and the rest of the information like description, tags, etc. For reference, the settings on the right hand side should match this issue, with the exception that the assignees on the PR are just those who are pushing code.
8. If you are the last person to push, release the pull request for reviews, and add @andrewdimmer as a reviewer.

**What to Add for each Class**

- The methods from the interface the class implements (if any); note these can be copy and pasted the interface diagram that was completed before
- Any private class variables (above the `__`); note there should be at least 1 as you need a reference to the next class in the chain
- A constructor that initializes all the private variables that don't have setters (below the `__` but above the `.. Use Case 1 ..`)
- Any private methods for Use Case 1 (below the public methods for Use Case 1)
 - Any private methods for Use Case 2 (below the public methods for Use Case 2)

Feel free to add as many or few private methods as you'd like to, this only changes your internal implementation and is entirely up to the team.

**Tasks to Complete**

- [x] Assign Tasks - @brendam19 , @juliawang2014 , or @martinhughes0722
- [x] Create Purpose Statement for Subteam - @martinhughes0722 
- [x] Create Purpose Statement for Production package - @martinhughes0722 
- [x] Create Purpose Statement for Helper package - @san4497 
- [x] Create Purpose Statement for Test package - @shayat97 
- [x] Create Purpose Statement for MiddlewareGisInterface - @BarczakS  
- [x] Create Purpose Statement for MiddlewareGisInterfaceImplementation - @duongy18418 
- [x] Create Purpose Statement for MiddlewareGisManager - @Slackjaw1431 
- [x] Create Purpose Statement for MiddlewareGisManagerImplementation - @juliawang2014 
- [x] Create Purpose Statement for MiddlewareGisProcessor - @juliawang2014 
- [x] Create Purpose Statement for MiddlewareGisProcessorImplementation - @juliawang2014 
- [x] etc.

**Additional context**
Note: if for some reason you need to change a public interface or shared data type, reach out to @andrewdimmer who can help you manage making that BREAKING CHANGE.

Note: if you need additional helper classes (and you likely will), just ping @andrewdimmer to get the file structure set up.

### PR DESCRIPTION
**Pull request by jvalentini555**
Close #40

### CODE REVIEW

File: docs/uml-class-diagrams/middleware01/production/MiddlewareGisManagerImplementation/MiddlewareGisManagerImplementation.puml [hunk 1]
```diff
@@ -2,11 +2,15 @@
 
 package edu.oakland.production.middleware01 {
     class MiddlewareGisManagerImplementation {
+        - MiddlewareGisProcessor middlewareGisProcessor
         __
+        + MiddlewareGisManagerImplementation(MiddlewareGisProcessor middlewareGisProcessor)
```
Comment (andrewdimmer) [inline docs/uml-class-diagrams/middleware01/production/MiddlewareGisManagerImplementation/MiddlewareGisManagerImplementation.puml]: This references database, right? If so, should it take a `DatabaseGisInterface` in the contructor?
```suggestion
        - DatabaseGisInterface databaseGisInterface
        __
        + MiddlewareGisManagerImplementation(DatabaseGisInterface databaseGisInterface)
```

**Review by andrewdimmer — CHANGES_REQUESTED** [review]
Still some work to do updating READMEs before this gets merged, but I wanted to point out a minor fix when I was looking through this: the `MiddlewareGisManager` calls to the `DatabaseGisInterface`, right?

File: docs/uml-class-diagrams/middleware01/production/MiddlewareGisManagerImplementation/MiddlewareGisManagerImplementation.puml (hunk 1 above)
Comment (jvalentini555) [reply docs/uml-class-diagrams/middleware01/production/MiddlewareGisManagerImplementation/MiddlewareGisManagerImplementation.puml]: Very good point, thanks for catching that!

File: docs/uml-class-diagrams/middleware01/production/MiddlewareGisManagerImplementation/MiddlewareGisManagerImplementation.puml [hunk 2]
```diff
@@ -2,12 +2,16 @@
 
 package edu.oakland.production.middleware01 {
     class MiddlewareGisManagerImplementation {
+        - DatabaseGisInterface databaseGisInterface
```
Comment (andrewdimmer) [inline docs/uml-class-diagrams/middleware01/production/MiddlewareGisManagerImplementation/MiddlewareGisManagerImplementation.puml:5]: You will probably need a private variable to stage the LocationDataPoint in between when you get it and when you get the signal from Middleware_02 to store it in the database.
```suggestion
        - DatabaseGisInterface databaseGisInterface
        - LocationDataPoint locationDataPoint
```

**Review by andrewdimmer — CHANGES_REQUESTED** [review]
Looks good so far! Just a few more purpose statements and one suggested private variable, then you should be good to go!

**Comment by jvalentini555** [comment]
@andrewdimmer I'm pretty sure I resolved the requests yet it still says there is a change requested. What am I missing?

**Review by andrewdimmer — APPROVED** [review]
Looks good to me! You are all set to merge! Well done!

**Review by camrongrant1 — APPROVED** [review]

**Review by brendam19 — APPROVED** [review]

**Pull request merged**